atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CQLKeyValueService.java

```diff
@@ -417,30 +423,44 @@ private void loadWithTs(final TableReference tableRef,
                             final ConsistencyLevel consistency) throws Exception {
-        final String loadWithTsQuery = "SELECT * FROM " + getFullTableName(tableRef) + " "
-                + "WHERE " + fieldName.row() + " = ? AND " + fieldName.column() + " = ? AND " + fieldName.timestamp()
-                + " > ?" + (!loadAllTs ? " LIMIT 1" : "");
         final CassandraKeyValueServiceConfig config = configManager.getConfig();
-        if (cells.size() > config.fetchBatchCount()) {
-            log.warn("A call to " + tableRef
-                    + " is performing a multiget " + cells.size()
-                    + " cells; this may indicate overly-large batching on a higher level.\n"
-                    + CassandraKeyValueServices.getFilteredStackTrace("com.palantir"));
-        }
-        final PreparedStatement preparedStatement = getPreparedStatement(tableRef, loadWithTsQuery, session).setConsistencyLevel(consistency);
-        List<ResultSetFuture> resultSetFutures = Lists.newArrayListWithCapacity(cells.size());
 
+        List<ResultSetFuture> resultSetFutures = Lists.newArrayListWithCapacity(cells.size());
+        TreeMultimap<byte[], Cell> cellsByCol =
+                TreeMultimap.create(UnsignedBytes.lexicographicalComparator(), Ordering.natural());
         for (Cell cell : cells) {
-            ResultSetFuture resultSetFuture = session.executeAsync(
-                    preparedStatement.bind(
-                            ByteBuffer.wrap(cell.getRowName()),
-                            ByteBuffer.wrap(cell.getColumnName()),
-                            ~startTs));
-            resultSetFutures.add(resultSetFuture);
+            cellsByCol.put(cell.getColumnName(), cell);
+        }
+        for (Entry<byte[], SortedSet<Cell>> entry : Multimaps.asMap(cellsByCol).entrySet()) {
+            if (entry.getValue().size() > config.fetchBatchCount()) {
+                log.warn("A call to " + tableRef
+                        + " is performing a multiget " + entry.getValue().size()
+                        + " cells; this may indicate overly-large batching on a higher level.\n"
+                        + CassandraKeyValueServices.getFilteredStackTrace("com.palantir"));
+            }
+            for (List<Cell> batch : Iterables.partition(entry.getValue(), config.fetchBatchCount())) {
+                String rowBinds = Joiner.on(",").join(Iterables.limit(Iterables.cycle('?'), batch.size()));
+                final String loadWithTsQuery = "SELECT * FROM " + getFullTableName(tableRef) + " "
+                        + "WHERE " + fieldName.row() + " IN (" + rowBinds + ") AND " + fieldName.column()
+                        + " = ? AND " + fieldName.timestamp() + " > ?" + (!loadAllTs ? " LIMIT 1" : "");
+                final PreparedStatement preparedStatement = getPreparedStatement(tableRef, loadWithTsQuery, session).setConsistencyLevel(consistency);
+                Object[] args = new Object[batch.size() + 2];
```

Comment:
this code is kind of brittle in that anyone who screws with it is likely to off by one...